### ISSUE TITLE
Clean up json key-value pair columns

### ISSUE THREAD

**Issue by jsonbecker** [tidyverse/tidyr#430]
It's fairly common to receive a CSV where one of the data elements is a character string which actually represents JSON text. This case comes up frequently when sourcing data from a NoSQL database or an RDBMS like PostreSQL that supports json/jsonb as a native column type.

In these scenarios, it'd be convenient to be able to transform the json text into a data.frame without using the other elements of the original CSV. In some ways, this is a classic unnest() problem where each value in a column actually represents multiple columns or rows of data.

This was first reported by @earino on Twitter:
https://twitter.com/earino/status/968544763430387712

Some complications to note: most JSON will not likely be a simple key-value pair set and contain several levels of nested attributes.

**Comment by jennybc** [comment]
I think there are several steps, some of which could potentially be shared with other tasks. Parsing the JSON column to a list column is step 1. Then this becomes an instance of the more general transpose or flatten problem. Basically you have the list you would have had if the JSON had come all at once vs. via csv export from a database.

Arguably, the thing that's unique to this situation is the need for a JSON vector parser ... in readr? This also has connections to the ongoing discussion of how/whether to support writing list-column-containing data frames to a plain text representation (vs., say, .rds). (https://github.com/tidyverse/readr/issues/303)

cc @jimhester @hadley

**Comment by hadley** [comment]
But we could have `col_json()`  (if storing as a column in a csv file is common)

**Comment by earino** [comment]
It's common enough that it happens to me regularly at more than one company. People store json objects in varchars now. Every database is schemaless.

**Comment by jennybc** [comment]
> I don't see the need for a parser in readr, isn't this just `lapply(df$json_column, jsonlite::fromJSON)` to parse and `vapply(df$json_column, jsonlite::toJSON, character(1))` for writing?

Yeah, it may very well be this simple, but it still might be nice to put a bow on it. I keep hearing about this type of data. And if we smoothed things out a little, it could also be the official solution to reading/writing list-columns.

**Comment by jimhester** [comment]
I JSON columns are quite common in databases, but I don't think terribly common in flat files like csvs and excel.

I agree having a function to make it easier to handle parsing this type of data makes sense, but not sure where the best place would be, and readr doesn't feel like a perfect fit.
